src/InsetOrderOptimizer.cpp

```diff
@@ -164,6 +168,55 @@ bool InsetOrderOptimizer::addToLayer()
     return added_something;
 }
 
+void InsetOrderOptimizer::insertSeamPoint(ExtrusionLine& closed_line)
+{
+    assert(closed_line.is_closed_);
+
+    Point2LL request_point;
+    switch (z_seam_config_.type_)
+    {
+    case EZSeamType::USER_SPECIFIED:
+        request_point = z_seam_config_.pos_;
+        break;
+    case EZSeamType::SHORTEST:
+        request_point = gcode_layer_.getLastPlannedPositionOrStartingPosition();
+        break;
+    default:
+        return;
+    }
+
+    // NOTE: Maybe rewrite this once we can use C++23 ranges::views::adjacent
+    size_t closest_junction_idx = 0;
+    coord_t closest_distance_sqd = std::numeric_limits<coord_t>::max();
+    for (const auto& [i, junction] : closed_line.junctions_ | ranges::views::enumerate)
+    {
+        const auto& next_junction = closed_line.junctions_[(i + 1) % closed_line.junctions_.size()];
+        const coord_t distance_sqd = LinearAlg2D::getDist2FromLineSegment(junction.p_, request_point, next_junction.p_);
+        if (distance_sqd < closest_distance_sqd)
+        {
+            closest_distance_sqd = distance_sqd;
+            closest_junction_idx = i;
+        }
+    }
+
+    const auto& start_pt = closed_line.junctions_[closest_junction_idx];
+    const auto& end_pt = closed_line.junctions_[(closest_junction_idx + 1) % closed_line.junctions_.size()];
+    const auto closest_point = LinearAlg2D::getClosestOnLineSegment(request_point, start_pt.p_, end_pt.p_);
+    constexpr coord_t smallest_dist_sqd = 25;
+    if (vSize2(closest_point - start_pt.p_) <= smallest_dist_sqd || vSize2(closest_point - end_pt.p_) <= smallest_dist_sqd)
+    {
+        return;
+    }
+
+    // NOTE: This could also be done on a single axis (skipping the implied sqrt), but figuring out which one and then using the right values became a bit messy/verbose.
+    const coord_t total_dist = vSize(end_pt.p_ - start_pt.p_);
+    const coord_t start_dist = vSize(closest_point - start_pt.p_);
+    const coord_t end_dist = vSize(closest_point - end_pt.p_);
+    const coord_t w = end_pt.w_ * end_dist / total_dist + start_pt.w_ * start_dist / total_dist;
```

Comment:
Could you add some `()` to this calculation ? That would help understanding.

Reply:
Yes, I can. Given my usual m.o. of just adding () everywhere I'm surprised that I even wrote a statement that long without _any_ 😅